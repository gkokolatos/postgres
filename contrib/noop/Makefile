# contrib/noop/Makefile

MODULE_big = noop
OBJS = noop.o

EXTENSION = noop
DATA = noop--1.0.sql
PGFILEDESC = "noop access method - informational NO-OP index"

REGRESS = noop

ifdef USE_PGXS
PG_CONFIG = pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
else
subdir = contrib/noop
top_builddir = ../..
include $(top_builddir)/src/Makefile.global
include $(top_srcdir)/contrib/contrib-global.mk
endif
