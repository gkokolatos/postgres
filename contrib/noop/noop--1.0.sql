/* contrib/noop/noop--1.0.sql */

-- complain if script is sourced in psql, rather than via CREATE EXTENSION
\echo Use "CREATE EXTENSION noop" to load this file. \quit

CREATE FUNCTION noophandler(internal)
RETURNS index_am_handler
AS 'MODULE_PATHNAME'
LANGUAGE C;

-- Access method
CREATE ACCESS METHOD noop TYPE INDEX HANDLER noophandler;
COMMENT ON ACCESS METHOD noop IS 'noop index access method';

-- Opclasses

CREATE OPERATOR CLASS int4_ops
DEFAULT FOR TYPE int4 USING noop AS
	OPERATOR	1	=(int4, int4),
	FUNCTION	1	hashint4(int4);

CREATE OPERATOR CLASS text_ops
DEFAULT FOR TYPE text USING noop AS
	OPERATOR	1	=(text, text),
	FUNCTION	1	hashtext(text);
