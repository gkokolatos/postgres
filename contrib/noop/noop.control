# noop extension
comment = 'noop access method - informational NO-OP index'
default_version = '1.0'
module_pathname = '$libdir/noop'
relocatable = false
