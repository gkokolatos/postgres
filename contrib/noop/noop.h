/*-------------------------------------------------------------------------
 *
 * noop.h
 *	  header file for greenplum noop access method implementation.
 *
 * IDENTIFICATION
 *	  contrib/noop/noop.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef NOOP_H
#define NOOP_H

#include "access/amapi.h"
#include "nodes/execnodes.h"
#include "fmgr.h"

extern Datum noophandler(PG_FUNCTION_ARGS);

/* NOOP index access method interface functions */
extern bytea *noopoptions(Datum reloptions, bool validate);
extern IndexBuildResult *noopbuild(Relation heap, Relation index,
				IndexInfo *indexInfo);
extern void noopbuildempty(Relation index);
extern void noopcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
				Cost *indexStartupCost, Cost *indexTotalCost,
				Selectivity *indexSelectivity, double *indexCorrelation);
extern bool noopinsert(Relation index, Datum *values, bool *isnull,
				ItemPointer ht_ctid, Relation heapRel,
				IndexUniqueCheck checkUnique);
extern IndexBulkDeleteResult *noopvacuumcleanup(IndexVacuumInfo *info,
				IndexBulkDeleteResult *stats);

#endif /* NOOP_H */
