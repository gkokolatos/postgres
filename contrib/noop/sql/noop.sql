CREATE EXTENSION noop;

CREATE TABLE noop_test (a integer, b integer, c text);
CREATE UNIQUE INDEX noop_test_b_key ON noop_test USING NOOP (b);
CREATE UNIQUE INDEX noop_test_c_key ON noop_test USING NOOP (c);
\d noop_test
\d noop_test_b_key
\d noop_test_c_key

SELECT
	c2.relname,
	i.indisprimary,
	i.indisunique,
	i.indisclustered,
	i.indisvalid,
	pg_catalog.pg_get_indexdef(i.indexrelid, 0, true),
	pg_catalog.pg_get_constraintdef(con.oid, true),
	contype,
	condeferrable,
	condeferred,
	i.indisreplident,
	c2.reltablespace
FROM
	pg_catalog.pg_class c,
	pg_catalog.pg_class c2,
	pg_catalog.pg_index i   LEFT JOIN
	pg_catalog.pg_constraint con
ON
(
	conrelid = i.indrelid AND
	conindid = i.indexrelid AND
	contype IN ('p','u','x')
)
WHERE
	(c2.relname = 'noop_test_b_key' OR
	c2.relname = 'noop_test_c_key') AND
	c.oid = i.indrelid AND
	i.indexrelid = c2.oid
ORDER BY
	i.indisprimary DESC,
	i.indisunique DESC,
	c2.relname
;

INSERT INTO noop_test (a, b, c) SELECT generate_series(0, 9) as a, 1 as b, 'one' as c;
SELECT count(b) as "b multiples", count(c) as "c multiples", b, c FROM noop_test GROUP BY b, c;

INSERT INTO noop_test SELECT i, i % 10, substr(md5(i::text), 1, 1) FROM generate_series(10,2009) i;
VACUUM ANALYZE noop_test;

DELETE FROM noop_test WHERE a %  3 = 0;
VACUUM FULL noop_test;

EXPLAIN (COSTS OFF) SELECT count(*) FROM noop_test WHERE b != 1;
EXPLAIN (COSTS OFF) SELECT count(*) FROM noop_test WHERE c = 'one';

SET enable_seqscan=off;
SET enable_bitmapscan=on;
SET enable_indexscan=on;

EXPLAIN (COSTS OFF) SELECT count(*) FROM noop_test WHERE b != 1;
EXPLAIN (COSTS OFF) SELECT count(*) FROM noop_test WHERE c = 'one';

RESET enable_seqscan;
RESET enable_bitmapscan;
RESET enable_indexscan;
