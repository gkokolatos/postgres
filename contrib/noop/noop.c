/*-------------------------------------------------------------------------
 *
 * noop.c
 *		Access methods for noop indexes.
 *
 * IDENTIFICATION
 *	  contrib/noop/noop.c
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "optimizer/cost.h"
#include "storage/bufmgr.h"
#include "utils/rel.h"
#include "utils/selfuncs.h"

#include "noop.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(noophandler);

/*
 * NOOP handler function:
 *     Creates and populates an IndexAmRoutine
 */
Datum
noophandler(PG_FUNCTION_ARGS)
{
	IndexAmRoutine *amroutine = makeNode(IndexAmRoutine);

	amroutine->amstrategies = 0;
	amroutine->amsupport = 1;
	amroutine->amcanorder = false;
	amroutine->amcanorderbyop = false;
	amroutine->amcanbackward = false;
	amroutine->amcanunique = true;
	amroutine->amcanmulticol = true;
	amroutine->amoptionalkey = true;
	amroutine->amsearcharray = false;
	amroutine->amsearchnulls = false;
	amroutine->amstorage = false;
	amroutine->amclusterable = false;
	amroutine->ampredlocks = false;
	amroutine->amkeytype = InvalidOid;

	amroutine->ambuild = noopbuild;
	amroutine->ambuildempty = noopbuildempty;
	amroutine->aminsert = noopinsert;
	amroutine->ambulkdelete = NULL;
	amroutine->amvacuumcleanup = noopvacuumcleanup;
	amroutine->amcanreturn = NULL;
	amroutine->amcostestimate = noopcostestimate;
	amroutine->amoptions = noopoptions;
	amroutine->amproperty = NULL;
	amroutine->amvalidate = NULL;
	amroutine->ambeginscan = NULL;
	amroutine->amrescan = NULL;
	amroutine->amgettuple = NULL;
	amroutine->amgetbitmap = NULL;
	amroutine->amendscan = NULL;
	amroutine->ammarkpos = NULL;
	amroutine->amrestrpos = NULL;

	PG_RETURN_POINTER(amroutine);
}

IndexBuildResult *
noopbuild(Relation heap, Relation index, IndexInfo *indexInfo)
{
	IndexBuildResult *result;

	/*
	 * This can never happen by us and indicates corruption
	 */
	if (RelationGetNumberOfBlocks(index) != 0)
		elog(ERROR, "index \"%s\" already contains data",
			 RelationGetRelationName(index));

	result = (IndexBuildResult *) palloc(sizeof(IndexBuildResult));
	result->heap_tuples = 0;
	result->index_tuples = 0;

	return result;
}

/*
 * Build an empty noop index in the initialization fork.
 */
void
noopbuildempty(Relation index)
{
	return;
}

bytea *
noopoptions(Datum reloptions, bool validate)
{
	return NULL;
}

/*
 * This function should not be called yet it needs to be defined since the index
 * am code asserts that it is present before it moves into an insertion.
 */
void
noopcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
			   Cost *indexStartupCost, Cost *indexTotalCost,
			   Selectivity *indexSelectivity, double *indexCorrelation)
{
	IndexOptInfo *index = path->indexinfo;
	List	   *qinfos;
	GenericCosts costs;

	/* Do preliminary analysis of indexquals */
	qinfos = deconstruct_indexquals(path);

	MemSet(&costs, 0, sizeof(costs));

	/* We have to visit all index tuples anyway */
	costs.numIndexTuples = index->tuples;

	/* Use generic estimate */
	genericcostestimate(root, path, loop_count, qinfos, &costs);

	*indexStartupCost = costs.indexStartupCost;
	*indexTotalCost = costs.indexTotalCost;
	*indexSelectivity = costs.indexSelectivity;
	*indexCorrelation = costs.indexCorrelation;

	return;
}

/*
 * Return value is always false as no value can ever be found.
 */
bool
noopinsert(Relation index, Datum *values, bool *isnull,
		 ItemPointer ht_ctid, Relation heapRel, IndexUniqueCheck checkUnique)
{
	return false;
}

/*
 * Post-VACUUM cleanup.
 *
 * Result: a palloc'd struct containing statistical info for VACUUM displays.
 */
IndexBulkDeleteResult *
noopvacuumcleanup(IndexVacuumInfo *info, IndexBulkDeleteResult *stats)
{

	if (stats == NULL)
		stats = (IndexBulkDeleteResult *) palloc0(sizeof(IndexBulkDeleteResult));

	return stats;
}
