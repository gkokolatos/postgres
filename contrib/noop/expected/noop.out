CREATE EXTENSION noop;
CREATE TABLE noop_test (a integer, b integer, c text);
CREATE UNIQUE INDEX noop_test_b_key ON noop_test USING NOOP (b);
CREATE UNIQUE INDEX noop_test_c_key ON noop_test USING NOOP (c);
\d noop_test
   Table "public.noop_test"
 Column |  Type   | Modifiers 
--------+---------+-----------
 a      | integer | 
 b      | integer | 
 c      | text    | 
Indexes:
    "noop_test_b_key" UNIQUE, noop (b)
    "noop_test_c_key" UNIQUE, noop (c)

\d noop_test_b_key
Index "public.noop_test_b_key"
 Column |  Type   | Definition 
--------+---------+------------
 b      | integer | b
unique, noop, for table "public.noop_test"

\d noop_test_c_key
Index "public.noop_test_c_key"
 Column | Type | Definition 
--------+------+------------
 c      | text | c
unique, noop, for table "public.noop_test"

SELECT
	c2.relname,
	i.indisprimary,
	i.indisunique,
	i.indisclustered,
	i.indisvalid,
	pg_catalog.pg_get_indexdef(i.indexrelid, 0, true),
	pg_catalog.pg_get_constraintdef(con.oid, true),
	contype,
	condeferrable,
	condeferred,
	i.indisreplident,
	c2.reltablespace
FROM
	pg_catalog.pg_class c,
	pg_catalog.pg_class c2,
	pg_catalog.pg_index i   LEFT JOIN
	pg_catalog.pg_constraint con
ON
(
	conrelid = i.indrelid AND
	conindid = i.indexrelid AND
	contype IN ('p','u','x')
)
WHERE
	(c2.relname = 'noop_test_b_key' OR
	c2.relname = 'noop_test_c_key') AND
	c.oid = i.indrelid AND
	i.indexrelid = c2.oid
ORDER BY
	i.indisprimary DESC,
	i.indisunique DESC,
	c2.relname
;
     relname     | indisprimary | indisunique | indisclustered | indisvalid |                         pg_get_indexdef                         | pg_get_constraintdef | contype | condeferrable | condeferred | indisreplident | reltablespace 
-----------------+--------------+-------------+----------------+------------+-----------------------------------------------------------------+----------------------+---------+---------------+-------------+----------------+---------------
 noop_test_b_key | f            | t           | f              | t          | CREATE UNIQUE INDEX noop_test_b_key ON noop_test USING noop (b) |                      |         |               |             | f              |             0
 noop_test_c_key | f            | t           | f              | t          | CREATE UNIQUE INDEX noop_test_c_key ON noop_test USING noop (c) |                      |         |               |             | f              |             0
(2 rows)

INSERT INTO noop_test (a, b, c) SELECT generate_series(0, 9) as a, 1 as b, 'one' as c;
SELECT count(b) as "b multiples", count(c) as "c multiples", b, c FROM noop_test GROUP BY b, c;
 b multiples | c multiples | b |  c  
-------------+-------------+---+-----
          10 |          10 | 1 | one
(1 row)

INSERT INTO noop_test SELECT i, i % 10, substr(md5(i::text), 1, 1) FROM generate_series(10,2009) i;
VACUUM ANALYZE noop_test;
DELETE FROM noop_test WHERE a %  3 = 0;
VACUUM FULL noop_test;
EXPLAIN (COSTS OFF) SELECT count(*) FROM noop_test WHERE b != 1;
         QUERY PLAN          
-----------------------------
 Aggregate
   ->  Seq Scan on noop_test
         Filter: (b <> 1)
(3 rows)

EXPLAIN (COSTS OFF) SELECT count(*) FROM noop_test WHERE c = 'one';
            QUERY PLAN             
-----------------------------------
 Aggregate
   ->  Seq Scan on noop_test
         Filter: (c = 'one'::text)
(3 rows)

SET enable_seqscan=off;
SET enable_bitmapscan=on;
SET enable_indexscan=on;
EXPLAIN (COSTS OFF) SELECT count(*) FROM noop_test WHERE b != 1;
         QUERY PLAN          
-----------------------------
 Aggregate
   ->  Seq Scan on noop_test
         Filter: (b <> 1)
(3 rows)

EXPLAIN (COSTS OFF) SELECT count(*) FROM noop_test WHERE c = 'one';
            QUERY PLAN             
-----------------------------------
 Aggregate
   ->  Seq Scan on noop_test
         Filter: (c = 'one'::text)
(3 rows)

RESET enable_seqscan;
RESET enable_bitmapscan;
RESET enable_indexscan;
